source/s2n/s2n_tls_channel_handler.c

```diff
@@ -558,7 +558,21 @@ static int s_s2n_handler_process_read_message(
 
         if (read < 0) {
             aws_mem_release(outgoing_read_message->allocator, outgoing_read_message);
-            continue;
+
+            /* the socket blocked so exit from the loop */
+            if (s2n_error_get_type(s2n_errno) == S2N_ERR_T_BLOCKED) {
```

Comment:
We do have to make an extra read to make sure that we have read everything.

> But maybe put back that && blocked == S2N_NOT_BLOCKED in the while loop, so we can break out of the loop earlier if we read > 0 data, and there's not more to read.

The blocked value is garbage if read > 0, so we cannot rely on that for any logic.

Reply:
I think we do have to make an extra read to make sure that we have read everything.

> But maybe put back that && blocked == S2N_NOT_BLOCKED in the while loop, so we can break out of the loop earlier if we read > 0 data, and there's not more to read.

The blocked value is garbage if `s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED`, and there will be no error if `read>0`.

Reply:
Looking at the s2n source code, `blocked` does appear to have meaning when `> 0` bytes are read.

at the [start of the function](https://github.com/aws/s2n-tls/blob/05ae36ce470f20233821f3cb000140124023b6bb/tls/s2n_recv.c#L137) it's initialized to:
`*blocked = S2N_BLOCKED_ON_READ;`

But then at the end, before returning the number of bytes processed, it [checks to see if there's more data available](https://github.com/aws/s2n-tls/blob/05ae36ce470f20233821f3cb000140124023b6bb/tls/s2n_recv.c#L231-L235):
```
    if (s2n_stuffer_data_available(&conn->in) == 0) {
        *blocked = S2N_NOT_BLOCKED;
    }

    return bytes_read;
```

Reply:
The way I understand it is that the issue with having that condition in the while loop for `read > 0` is as follows: it tries to see if we have read everything. If yes, then it sets the status to NOT_BLOCKED, and in the next call, because `blocked == NOT_BLOCKED`, it will return that there was no data available. However, if the buffer was full and there is still data available, the status would be blocked, and we will never read the remaining data.

Reply:
oh yeah, you're right.
we want to bail out of the loop if read > 0 and NOT_BLOCKED